lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,631 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(zzhc, CONFIG_ZZHC_LOG_LEVEL);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <net/socket.h>
+#include <lte_lc.h>
+#include "zzhc_internal.h"
+
+#define REGVER            2         /** Self-registration protocol version */
+#define MANUFACTURER_CODE "NOD"     /** Manufactuerer code */
+#define SERVER_PORT       9999      /** Server's port number */
+#define RX_BUF_LEN        128       /** Length of Rx Buffer */
+#define JSON_MAX_LEN      320       /** Length of json */
+#define IMEI_LEN          15        /** Length of IMEI */
+#define RETRY_TIMEOUT     3600      /** 1 hour, 3600s */
+#define RETRY_CNT_MAX     3         /** Max. value of retry counter */
+
+#define THREAD_PRIORITY   K_PRIO_PREEMPT(CONFIG_ZZHC_THREAD_PRIO)
+#define STACK_SIZE        CONFIG_ZZHC_STACK_SIZE
+
+struct plmn {
+	unsigned short mcc;
+	unsigned short mnc;
+};
+
+/**@brief Host name of self-registration server. */
+static char const hostname[] = "zzhc.vnet.cn";
+
+/**@brief +CEREG event prefix. */
+static char const status_cereg[7] = "+CEREG:";
+
+/**@brief Lookup table for chip revision. */
+static char const chip_rev[4] = "XA01";
+
+/**@brief PLMN whitelist. */
+static struct plmn const simcard_whitelist[] = {
+	{0x454, 0x031},         /**< MCC=454, MNC=31, China Telecom Global Ltd*/
+	{0x460, 0x003},         /**< MCC=460, MNC=03, China Telecom */
+	{0x460, 0x005},         /**< MCC=460, MNC=05, China Telecom */
+	{0x460, 0x011},         /**< MCC=460, MNC=11, China Telecom */
+	{0x455, 0x002},         /**< MCC=455, MNC=02, China Telecom (Macau) */
+	{0x455, 0x007},         /**< MCC=455, MNC=07, China Telecom (Macau) */
+};
+
+/**@brief Function to handle AT-command notifications. */
+static void at_notif_handler(void *context, char *response)
+{
+	int rc;
+	struct zzhc *ctx = (struct zzhc *)context;
+
+	if (ctx == NULL) {
+		return;
+	}
+
+	if (memcmp(response, status_cereg, sizeof(status_cereg)) == 0) {
+		rc = zzhc_get_at_param_short(ctx, response, 0);
+		if (rc < 0) {
+			return;
+		}
+
+		LOG_DBG("+CEREG: %d", rc);
+		if (rc == LTE_LC_NW_REG_REGISTERED_HOME
+			|| rc == LTE_LC_NW_REG_REGISTERED_ROAMING) {
```

Comment:
nit: Could possibly keep the `||` operator in the line above, for consistency.

Reply:
Will be changed accordingly in the next push.

Reply:
Done in the last push on 7th Nov, 2019 (ref: [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1340#event-2777928390)).